stdlib/public/core/DiscontiguousSlice.swift

```diff
@@ -0,0 +1,451 @@
+/// indexed by a set of indices.
+@available(SwiftStdlib 5.11, *)
+public struct DiscontiguousSlice<Base: Collection> {
+  internal var _base: Base
+  
+  /// The set of subranges that are available through this discontiguous slice.
+  public let subranges: RangeSet<Base.Index>
+
+  /// The collection that the indexed collection wraps.
+  public var base: Base {
+    return _base
+  }
+
+  internal init(_base: Base, subranges: RangeSet<Base.Index>) {
+    self._base = _base
+    self.subranges = subranges
+  }
+}
+
+@available(SwiftStdlib 5.11, *)
+extension DiscontiguousSlice: Sendable
+where Base: Sendable, Base.Index: Sendable {}
+
+@available(SwiftStdlib 5.11, *)
+extension DiscontiguousSlice: Equatable where Base.Element: Equatable {
+  public static func ==(lhs: Self, rhs: Self) -> Bool {
+    lhs.elementsEqual(rhs)
+  }
+}
+
+@available(SwiftStdlib 5.11, *)
+extension DiscontiguousSlice: Hashable where Base.Element: Hashable {
+  public func hash(into hasher: inout Hasher) {
+    var count = 0
+    for element in self {
+      hasher.combine(element)
+      count += 1
+    }
+    hasher.combine(count) // discriminator
+  }
+}
+
+@available(SwiftStdlib 5.11, *)
+extension DiscontiguousSlice: CustomStringConvertible {
+  public var description: String {
+    _makeCollectionDescription()
+  }
+}
+
+@available(SwiftStdlib 5.11, *)
+extension DiscontiguousSlice {
+  /// A position in a `DiscontiguousSlice`.
+  public struct Index {
+    /// The index of the range that contains `base`.
+    internal let _rangeOffset: Int
+    
+    /// The position of this index in the base collection.
+    public let base: Base.Index
+
+    internal init(_rangeOffset: Int, base: Base.Index) {
+      self._rangeOffset = _rangeOffset
+      self.base = base
+    }
+  }
+}
+
+@available(SwiftStdlib 5.11, *)
+extension DiscontiguousSlice.Index: Equatable {
+  public static func == (left: Self, right: Self) -> Bool {
+    left.base == right.base
+  }
+}
+
+@available(SwiftStdlib 5.11, *)
+extension DiscontiguousSlice.Index: Hashable where Base.Index: Hashable {
+  public func hash(into hasher: inout Hasher) {
+    hasher.combine(base)
+  }
+}
+
+@available(SwiftStdlib 5.11, *)
+extension DiscontiguousSlice.Index: Comparable {
+  public static func < (left: Self, right: Self) -> Bool {
+    left.base < right.base
+  }
+}
+
+@available(SwiftStdlib 5.11, *)
+extension DiscontiguousSlice.Index: CustomStringConvertible {
+  public var description: String {
+    var d = ""
+    debugPrint(base, terminator: "", to: &d)
+    d += " (rangeOffset: \(_rangeOffset))"
+    return d
+  }
+}
+
+@available(SwiftStdlib 5.11, *)
+extension DiscontiguousSlice.Index: Sendable where Base.Index: Sendable {}
+
+@available(SwiftStdlib 5.11, *)
+extension DiscontiguousSlice: Sequence {
+  public typealias Element = Base.Element
+  public typealias Iterator = IndexingIterator<Self>
+
+  public func _customContainsEquatableElement(_ element: Element) -> Bool? {
```

Comment:
Are these hidden customization points exercised by the new tests?

(This applies to all public-but-underscored members that implement protocol requirements, both on `DiscontiguousSlice` and `RangeSet`.)

Reply:
I've added tests to cover all of the public, underscored members